server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -129,8 +137,23 @@ public void checkIndexSplitAllowed() {}
 
         protected abstract int shardId(String id, @Nullable String routing);
 
+        @Override
+        public void process(IndexRequest indexRequest) {
```

Comment:
I'd resisted adding `IndexRequest` to this class in the past. I think I can live with it - at least for this. Maybe for all the other routing methods in a follow up change. Not sure.